### PR TITLE
Added skeleton of Query layer.

### PR DESCRIPTION
**Pull request by Bplotka**
TODO:
* [x] Lock working dependencies

In some portion fulfills issue: https://github.com/improbable-eng/promlts/issues/6

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by Bplotka** [comment]
image

**Comment by Bplotka** [comment]
Main exploding subpackages of Prometheus are: 
`github.com/prometheus/prometheus/discovery` (imported by `github.com/prometheus/prometheus/retrieval` we are directly importing)
Also a little bit: (e.g unnecessary aws package)
`github.com/prometheus/prometheus/config` (`direct import)

**Comment by Bplotka** [comment]
OK, I am assuming there is no way of avoiding this without reimplementing "github.com/prometheus/prometheus/web/api/v1"... so I am going to try to find versions that does not collide.

**Comment by Bplotka** [comment]
One deps mismatch fixed: https://github.com/kubernetes/client-go/issues/312

**Comment by fabxc** [comment]
Yes, it gets kind of messy. Decoupling the integrations from the `retrieval` package is kind of a long-standing issues. I believe someone has actually filed an upstream PR finally but I haven't gotten around to reviewing it yet.

FWIW, the query API alone is pretty trivial. Would it be that bad to just reimplement it inline in this repo and copy a few code fragments here and there?
We would basically never have to touch it again?

File: pkg/query/compression/compression.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package compression
```
Comment (fabxc) [inline pkg/query/compression/compression.go]: Do we need this? IIRC the standard library negotiates and implements gzipping by default nowadays.

Comment (Bplotka) [reply pkg/query/compression/compression.go]: yes, this might be redundant, will kill it

File: pkg/query/api.go [hunk 2]
```diff
@@ -0,0 +1,445 @@
+package query
```
Comment (fabxc) [inline pkg/query/api.go]: Probably good to have some sort of notice on this file that this is largely copied from prometheus/prometheus and the respective licensing applies. License header needs to be included I think.

File: cmd/promlts/query.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	reg prometheus.Registerer,
+	apiAddr string,
+	cfg query.Config,
+) error {
+	level.Info(logger).Log(
+		"msg", "I'm a query node",
+		"api-address", apiAddr,
+		"store.addresses", cfg.StoreAddresses,
+		"query.timeout", cfg.QueryTimeout,
+		"query.max-concurrent", cfg.MaxConcurrentQueries,
```
Comment (fabxc) [inline cmd/promlts/query.go]: We should probably just have a simple message for each command that it started. Probably we should move them right before `g.Run` as everything before is setup.

File: cmd/promlts/query.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+// registerQuery registers a query command.
+func registerQuery(app *kingpin.Application, name string) runFunc {
+	cmd := app.Command(name, "query node exposing PromQL enabled Query API with data retrieved from multiple store nodes")
+
+	apiAddr := cmd.Flag("api-address", "listen address for the query API").
+		Default(":19099").String()
+
+	storeAddresses := cmd.Flag("store.addresses", "comma delimited listen addresses of store APIs").
+		Default("localhost:19090").Strings()
```
Comment (fabxc) [inline cmd/promlts/query.go]: I was wondering how we should go about discovery. Probably we'll want to support a plain list, DNS A + port, and DNS SRV. But we can add those capabilities later.

File: cmd/promlts/query.go (hunk 3 above)
Comment (Bplotka) [reply cmd/promlts/query.go]: Well, is it bad to have few more arguments with config values? I will move that after the `group` setup.

File: pkg/query/api.go (hunk 2 above)
Comment (Bplotka) [reply pkg/query/api.go]: I added that notice to the API struct comment, but can do that below header

**Comment by fabxc** [comment]
👍

**Pull request merged**